From b866df61739193ebf3b8ece81774033d450d9498 Mon Sep 17 00:00:00 2001
From: Tony Wickham <twickham@google.com>
Date: Fri, 22 Aug 2025 12:10:30 -0700
Subject: [PATCH 22/56] Add onKeyEvent, to support KEYCODE_HOME

- Also added displayId to onBackEvent

Bug: 381339822
Flag: com.android.launcher3.home_button_uses_keycode_home
Test: Manual, OverviewCommandHelperTest
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:622a882e1f17d108a336ecfd3e4eb7aa9707f1e6)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:0f46d701bb4d6d81106e9a5a0a13e7237d311efa
Merged-In: I55e2ee7adfa40c9d8f289a2b4c76aa12f80e78c3
Change-Id: I55e2ee7adfa40c9d8f289a2b4c76aa12f80e78c3
---
 .../systemui/shared/recents/ISystemUiProxy.aidl | 12 ++++++++++--
 .../android/systemui/LauncherProxyService.java  | 17 +++++++++++------
 2 files changed, 21 insertions(+), 8 deletions(-)

diff --git a/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl b/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
index 4b1e52be5..516e01c0c 100644
--- a/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
+++ b/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
@@ -103,9 +103,10 @@ interface ISystemUiProxy {
     oneway void expandNotificationPanel() = 29;
 
     /**
-     * Notifies SystemUI of a back KeyEvent.
+     * Notifies SystemUI of KEYCODE_BACK. If the passed KeyEvent is not null, SystemUI may use it to
+     * show a predictive back animation, otherwise it will send a ACTION_DOWN followed by ACTION_UP.
      */
-    oneway void onBackEvent(in KeyEvent keyEvent) = 44;
+    oneway void onBackEvent(in KeyEvent keyEvent, int displayId) = 44;
 
     /** Sets home rotation enabled. */
     oneway void setHomeRotationEnabled(boolean enabled) = 45;
@@ -190,4 +191,11 @@ interface ISystemUiProxy {
     oneway void notifyRecentsButtonPositionChanged(in Rect position) = 59;
 
     // Next id = 60
+
+    /**
+     * Notifies SystemUI of a KeyEvent of the specified type (e.g. KEYCODE_BACK, KEYCODE_HOME).
+     */
+    oneway void onKeyEvent(int keycode, int displayId) = 60;
+
+    // Next id = 61
 }
diff --git a/packages/SystemUI/src/com/android/systemui/LauncherProxyService.java b/packages/SystemUI/src/com/android/systemui/LauncherProxyService.java
index 669ff8de3..ab57dfe31 100644
--- a/packages/SystemUI/src/com/android/systemui/LauncherProxyService.java
+++ b/packages/SystemUI/src/com/android/systemui/LauncherProxyService.java
@@ -345,21 +345,26 @@ public class LauncherProxyService implements CallbackController<LauncherProxyLis
         }
 
         @Override
-        public void onBackEvent(@Nullable KeyEvent keyEvent) throws RemoteException {
-            final int displayId = keyEvent == null ? INVALID_DISPLAY : keyEvent.getDisplayId();
+        public void onBackEvent(@Nullable KeyEvent keyEvent, int displayId) throws RemoteException {
             if (predictiveBackThreeButtonNav() && predictiveBackSwipeEdgeNoneApi()
                     && mBackAnimation != null && keyEvent != null) {
                 mBackAnimation.setTriggerBack(!keyEvent.isCanceled());
                 mBackAnimation.onBackMotion(/* touchX */ 0, /* touchY */ 0, keyEvent.getAction(),
                         EDGE_NONE, displayId);
             } else {
-                verifyCallerAndClearCallingIdentityPostMain("onBackPressed", () -> {
-                    sendEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_BACK, displayId);
-                    sendEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_BACK, displayId);
-                });
+                onKeyEvent(KeyEvent.KEYCODE_BACK, displayId);
             }
         }
 
+        public void onKeyEvent(int keycode, int displayId) {
+            verifyCallerAndClearCallingIdentityPostMain(
+                    "onKeyEvent " + KeyEvent.keyCodeToString(keycode) + " displayId=" + displayId,
+                    () -> {
+                        sendEvent(KeyEvent.ACTION_DOWN, keycode, displayId);
+                        sendEvent(KeyEvent.ACTION_UP, keycode, displayId);
+                    });
+        }
+
         @Override
         public void injectLongPress(int keyCode) throws RemoteException {
             final int displayId = mContext.getDisplayId();
-- 
2.52.0

